transports/quic/src/transport.rs

```diff
@@ -0,0 +1,707 @@
+                        error: err.into(),
+                    })
+                }
+            }
+        }
+    }
+
+    /// Poll [`DialerState`] to initiate requested dials.
+    fn poll_dialer(&mut self, cx: &mut Context<'_>) -> Poll<TransportError> {
+        let Self {
+            dialer_state,
+            endpoint_channel,
+            ..
+        } = &mut *self;
+
+        dialer_state.poll(endpoint_channel, cx)
+    }
+}
+
+impl Stream for Listener {
+    type Item = TransportEvent<Connecting, TransportError>;
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            if let Some(event) = self.pending_event.take() {
+                return Poll::Ready(Some(event));
+            }
+            if self.is_closed {
+                return Poll::Ready(None);
+            }
+            match self.poll_if_addr(cx) {
+                Poll::Ready(event) => return Poll::Ready(Some(event)),
+                Poll::Pending => {}
+            }
+            match self.poll_dialer(cx) {
+                Poll::Ready(error) => {
+                    self.close(Err(error));
+                    continue;
+                }
+                Poll::Pending => {}
+            }
+            match self.new_connections_rx.poll_next_unpin(cx) {
+                Poll::Ready(Some(connection)) => {
+                    let local_addr = socketaddr_to_multiaddr(connection.local_addr());
+                    let send_back_addr = socketaddr_to_multiaddr(&connection.remote_addr());
+                    let event = TransportEvent::Incoming {
+                        upgrade: Connecting::from_connection(connection),
+                        local_addr,
+                        send_back_addr,
+                        listener_id: self.listener_id,
+                    };
+                    return Poll::Ready(Some(event));
+                }
+                Poll::Ready(None) => {
+                    self.close(Err(TransportError::EndpointDriverCrashed));
+                    continue;
+                }
+                Poll::Pending => {}
+            };
+            self.waker = Some(cx.waker().clone());
+            return Poll::Pending;
+        }
+    }
+}
+
+impl Drop for Listener {
+    fn drop(&mut self) {
+        self.endpoint_channel.send_on_drop(ToEndpoint::Decoupled);
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum SocketFamily {
+    Ipv4,
+    Ipv6,
+}
+
+impl SocketFamily {
+    fn is_same(a: &IpAddr, b: &IpAddr) -> bool {
+        matches!(
+            (a, b),
+            (IpAddr::V4(_), IpAddr::V4(_)) | (IpAddr::V6(_), IpAddr::V6(_))
+        )
+    }
+}
+
+impl From<IpAddr> for SocketFamily {
+    fn from(ip: IpAddr) -> Self {
+        match ip {
+            IpAddr::V4(_) => SocketFamily::Ipv4,
+            IpAddr::V6(_) => SocketFamily::Ipv6,
+        }
+    }
+}
+
+/// Turn an [`IpAddr`] into a listen-address for the endpoint.
```

Comment:
Maybe the doc comment should mention that this merged the endpoint addr port with the ip (overriding any port at the ip). Took me a bit to understand what this was doing and why.

Reply:
Doc added in 3fa3e9ef635b991a14d49898fd72570eea612594.

> overriding any port at the ip

Not sure I understand what you mean. It does not overwrite anything in the ip, but instead combines ip and port.

Reply:
sorry, I must have missed that this just provides the ip address no port. Docs look good